### PR TITLE
feat(terraform): Add hyphen and non-ASCII support for domain names in credential extraction

### LINKED ISSUE #6068: feat(terraform): hyphen and non-ASCII support for domain names in credential extraction
### Discussed in https://github.com/aquasecurity/trivy/discussions/6067

<div type='discussions-op-text'>

<sup>Originally posted by **adam-carruthers** February  5, 2024</sup>
### Description

Here is the trivy code that finds the token for private terraform registry modules:

```go
envVar := fmt.Sprintf("TF_TOKEN_%s", strings.ReplaceAll(hostname, ".", "_"))
```

https://github.com/aquasecurity/trivy-iac/blob/3d8e412cf0ab0c8ad0296df8a01363a3ed7efe85/pkg/scanners/terraform/parser/resolvers/registry.go#L58

This correctly changes the `.` characters in hostnames to `_` - but that's not the only transformation that needs to be done. It also needs to look for a second env var with the `-` characters as `__`. From the terraform docs:

> Hyphens are also valid within host names but usually invalid as variable names and may be encoded as double underscores. For example, you can set a token for the domain name café.fr as TF_TOKEN_xn--caf-dma.fr, TF_TOKEN_xn--caf-dma_fr, or TF_TOKEN_xn____caf__dma_fr. If multiple variables evaluate to the same hostname, Terraform will choose the one defined last in the operating system's variable table.
>
> https://developer.hashicorp.com/terraform/cli/config/config-file

I really need this change, because I pass the token through the `TF_TOKEN` variable, but the hostname has a `-` and I am using Ubuntu that doesn't allow `-` in environment variables, so trivy can't find my 

### Desired Behavior

When I used my environment variable `TF_TOKEN_some__site_with_hash_com` trivy recognised it as the terraform private registry token for `some-site.with.hash.com`

### Actual Behavior

Trivy didn't recognise my environment variable and I couldn't connect it to the private registry.

### Reproduction Steps

1. Add terraform code to a folder that uses some imaginary remote repository (doesn't have to be real)

```terraform
module "fake" {
  source = "some-site.with.hash.com/something/foobar/aws"
}
```

2. Add the environment variable in your bash terminal

```bash
export TF_TOKEN_some__site_with_hash_com=abcdef
```

3. Run trivy in debug mode

```bash
trivy config --skip-dirs .terraform -d .
```

4. Observe the debug output that it didn't find the token

### Target

Filesystem

### Scanner

Misconfiguration

### Output Format

None

### Mode

Standalone

### Debug Output

```bash
2024-02-05T13:59:27.594Z        DEBUG   [misconf] 59:27.594451288 terraform.parser.<root>.evaluator.resolver Resolving module 'module.fake' with source: 'some-site.with.hash.com/something/foobar/aws'...
2024-02-05T13:59:27.594Z        DEBUG   [misconf] 59:27.594463619 terraform.parser.<root>.evaluator.resolver Trying to resolve: 4f54644b2a174fe837cdc15cfeb7b7ba
2024-02-05T13:59:27.594Z        DEBUG   [misconf] 59:27.594532381 terraform.parser.<root>.evaluator.resolver No token was found for the registry at some-site.with.hash.com
```

### Operating System

Ubuntu

### Version

```bash
Version: 0.49.0
Policy Bundle:
  Digest: sha256:f21e8e92a7b3f6042ef7acfd3b799afc1648536dc4111c4d5458bc16396f8332
  DownloadedAt: 2024-02-05 09:59:15.002900989 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)</div>

### PR DESCRIPTION
**Pull request by adam-carruthers**
## Related issues
- Close #6068 

## Related PRs
- [x] https://github.com/aquasecurity/trivy-iac/pull/95 - This PR was open and almost accepted in trivy-iac, but that repo was merged into the main trivy repo very recently

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [x] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [x] I've added usage information (if the PR introduces new options) - **NA**
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change). - **NA**

### CODE REVIEW

File: pkg/iac/scanners/terraform/parser/resolvers/registry.go [hunk 1]
```diff
@@ -136,6 +136,29 @@ func (r *registryResolver) Resolve(ctx context.Context, target fs.FS, opt Option
...
+		}
+
+		if token == "" {
+			return "", fmt.Errorf("No token was found for the registry at %s", hostname)
+		} else {
+			return token, nil
+		}
+	}
+}
+
```
Comment (simar7) [inline pkg/iac/scanners/terraform/parser/resolvers/registry.go:161]: How about if we drop the trailing else?

```suggestion
func getPrivateRegistryTokenFromEnvVars(hostname string) (string, error) {
	token := ""
	asciiHostname, err := idna.ToASCII(hostname)
	if err != nil {
		return "", fmt.Errorf("Could not convert hostname %s to a punycode encoded ASCII string so cannot find token for this registry", hostname)
	}

	envVar := fmt.Sprintf("TF_TOKEN_%s", strings.ReplaceAll(asciiHostname, ".", "_"))
	token = os.Getenv(envVar)

	// Dashes in the hostname can optionally be converted to double underscores
	if token == "" {
		envVar = strings.ReplaceAll(envVar, "-", "__")
		token = os.Getenv(envVar)
	}

	if token == "" {
		return "", fmt.Errorf("No token was found for the registry at %s", hostname)
	}
	return token, nil
	
}

```

**Review by simar7 — APPROVED** [review]
Thanks for recreating the PR. It lgtm, just left one small comment to improve readability.

File: pkg/iac/scanners/terraform/parser/resolvers/registry_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
+package resolvers
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_getPrivateRegistryTokenFromEnvVars(t *testing.T) {
```
Comment (nikpivkin) [inline pkg/iac/scanners/terraform/parser/resolvers/registry_test.go:9]: How about using a table test here? Here's an example https://github.com/aquasecurity/trivy/blob/main/pkg/licensing/classifier_test.go#L15

File: pkg/iac/scanners/terraform/parser/resolvers/registry.go (hunk 1 above)
Comment (adam-carruthers) [reply pkg/iac/scanners/terraform/parser/resolvers/registry.go]: Agree, added

File: pkg/iac/scanners/terraform/parser/resolvers/registry_test.go (hunk 2 above)
Comment (adam-carruthers) [reply pkg/iac/scanners/terraform/parser/resolvers/registry_test.go]: Done, see what you think

**Comment by adam-carruthers** [comment]
Sorry, was on holiday! Sorry it took me a little bit to get back to it.

**Review by simar7 — APPROVED** [review]

**Pull request merged**